### DISCUSSION TITLE
[noob] Failing to install via pipx in Docker

### DISCUSSION THREAD

**Discussion by rieje** [caksoylar/keymap-drawer#108] [Q&A] [answered]
I get the following error trying to install keymap-drawer via pipx in a [Docker container](https://github.com/KinesisCorporation/Adv360-Pro-ZMK/blob/V3.0/Dockerfile). In the container, I manually installed `pipx python3-venv python3` and then run `pipx install keymap-drawer`, which results in the following error: 

> ERROR: Ignored the following versions that require a different python version: 0.1.0 Requires-Python >=3.10,<4.0; 0.1.1 Requires-Python >=3.10,<4.0; 0.1.2 Requires-Python >=3.10,<4.0; 0.1.3 Requires-Python >=3.10,<4.0; 0.10.0 Requires-Python >=3.10,<4.0; 0.11.0 Requires-Python >=3.10,<4.0; 0.12.0 Requires-Python >=3.10,<4.0; 0.12.1 Requires-Python >=3.10,<4.0; 0.13.0 Requires-Python >=3.10,<4.0; 0.13.1 Requires-Python >=3.10,<4.0; 0.13.2 Requires-Python >=3.10,<4.0; 0.13.3 Requires-Python >=3.10,<4.0; 0.13.4 Requires-Python >=3.10,<4.0; 0.14.0 Requires-Python <4.0,>=3.10; 0.14.1 Requires-Python <4.0,>=3.10; 0.15.0 Requires-Python <4.0,>=3.10; 0.16.0 Requires-Python <4.0,>=3.10; 0.17.0 Requires-Python <4.0,>=3.10; 0.2.0 Requires-Python >=3.10,<4.0; 0.3.0 Requires-Python >=3.10,<4.0; 0.4.0 Requires-Python >=3.10,<4.0; 0.5.0 Requires-Python >=3.10,<4.0; 0.6.0 Requires-Python >=3.10,<4.0; 0.7.0 Requires-Python >=3.10,<4.0; 0.7.1 Requires-Python >=3.10,<4.0; 0.8.0 Requires-Python >=3.10,<4.0; 0.9.0 Requires-Python >=3.10,<4.0
ERROR: Could not find a version that satisfies the requirement keymap-drawer (from versions: none)
ERROR: No matching distribution found for keymap-drawer
'/root/.local/pipx/venvs/keymap-drawer/bin/python -m pip install keymap-drawer -q' failed

But `python3 --version` shows `Python 3.8.10`.

Any ideas? I want to get this installed before before trying building from source for the merged but not yet released [zmk_additional_includes](https://github.com/caksoylar/keymap-drawer/pull/105).

I was thinking of manually running keymap-drawer on the host system instead, but I'm using zmk-helpers (see PR above) and the west toolchain to build ZMK firmware pulls it from the container for building so I don't actually need a local copy of zmk-helpers (only my zmk-config which is mounted as a Docker volume for the container). If I were to use `zmk-helpers` on host system instead of container, it looks like I need a copy of `zmk-helpers` just for this.

**Comment by caksoylar** [comment]
It looks like the problem is Python version, your container has 3.8 and KD requires 3.10. But the image is derived from Ubuntu Noble https://github.com/zmkfirmware/zmk-docker/blob/3.5-branch/Dockerfile#L1 and that seems to have 3.12 for `python3`: https://packages.ubuntu.com/noble/amd64/python3. So I am not sure why that would be the case. I assume `python --version` reports the same?

Does anything change if you spin up a fresh container from the same Dockerfile?

> I was thinking of manually running keymap-drawer on the host system instead, but I'm using zmk-helpers (see PR above) and the west toolchain to build ZMK firmware pulls it from the container for building so I don't actually need a local copy of zmk-helpers (only my zmk-config which is mounted as a Docker volume for the container). If I were to use zmk-helpers on host system instead of container, it looks like I need a copy of zmk-helpers just for this.

Can you not mount the project folder (that will contain a copy of `zmk-helpers` after `west update`) as a volume to access it from the host? This way you could run KD on the host and point to the shared volume.

**Reply by rieje** [reply]
> It looks like the problem is Python version, your container has 3.8 and KD requires 3.10. But the image is derived from Ubuntu Noble https://github.com/zmkfirmware/zmk-docker/blob/3.5-branch/Dockerfile#L1 and that seems to have 3.12 for `python3`: https://packages.ubuntu.com/noble/amd64/python3. So I am not sure why that would be the case. I assume `python --version` reports the same?

In the Dockerfile I added `RUN apt update && apt install -y pipx python3-venv python3`--if I don't update the repo, Ubuntu's `apt` throws the error:

> Package python3-venv is not available, but is referred to by another package.
This may mean that the package is missing, has been obsoleted, or
is only available from another source
E: Unable to locate package pipx
E: Package 'python3-venv' has no installation candidate

I didn't install python so no output for `python --version`, but I installed `python-is-python3` and then `python --version` outputs the same as `python3 --version` but same results. `pipx` is version 0.12.3.1. 

> Does anything change if you spin up a fresh container from the same Dockerfile?

I repeated the above steps (removing the stale image and starting new container) but same results. The Dockerfile can just contain that `RUN` layer and the `Makefile` uses Docker like this:

      docker buildx build --ssh=default --add-host=$(GIT_SERVER):$(SERVER_IP) \
        --tag zmk --build-arg=GIT_SERVER=$(GIT_SERVER) \
        --build-arg=SSH_PORT=$(SSH_PORT) --file Dockerfile .
      docker run --rm -it --name zmk \
        -v $(PWD)/firmware:/app/firmware$(SELINUX1) \
        -v $(PWD)/zmk-config/config:/app/config:ro$(SELINUX2) \
        -e TIMESTAMP=$(TIMESTAMP) \
        -e COMMIT=$(COMMIT) \
        -e BUILD_RIGHT=true \
        zmk bash        # bash for debugging

**Reply by caksoylar** [reply, marked as answer]
I tried installing myself and checking things out. The issue is the `stable` tag of the image is [still Ubuntu 20.04](https://hub.docker.com/layers/zmkfirmware/zmk-build-arm/stable/images/sha256-a2c1634e7a0a5e103124010235557f2176d81ecd23937bc23067bedeb7a16bf2?context=explore), so it only has Python 3.8.

If you use `docker.io/zmkfirmware/zmk-build-arm:3.5-branch` instead, it will use the latest zmk-docker image with Ubuntu 24.04 and that comes with Python 3.12 already with the `python3` command. I don't think you should worry about using the updated image either, ZMK builds should continue to work fine with it.

**Reply by caksoylar** [reply]
@rieje FYI the `stable` tag was just bumped, so you can switch back to that if you like.

**Comment by caksoylar** [comment]
> Any ideas? I want to get this installed before before trying building from source for the merged but not yet released https://github.com/caksoylar/keymap-drawer/pull/105.

As an FYI, once you get `pip/pipx` install working you can install from source using the git repo url:

```sh
pipx install keymap-drawer@git+https://github.com/caksoylar/keymap-drawer.git
```
